### PR TITLE
[wip] type refactoring for abi

### PR DESCRIPTION
**Pull request by dark64**
- [x] Tests should fail as something broke within semantics
- [x] Add tests for abi serialization

### CODE REVIEW

File: zokrates_core/src/static_analysis/propagation.rs [hunk 1]
```diff
@@ -433,7 +445,13 @@ impl<'ast, T: Field> Folder<'ast, T> for Propagator<'ast, T> {
 
                 match s.into_inner() {
                     StructExpressionInner::Value(v) => {
-                        match members.iter().zip(v).find(|(id, _)| id.0 == m).unwrap().1 {
+                        match members
```
[the remark this answers is not in the document]
Comment (dark64) [reply zokrates_core/src/static_analysis/propagation.rs]: I think this is correct, second member is used for matching after unwrapping? `unwrap().1`

File: zokrates_abi/src/lib.rs [hunk 2]
```diff
@@ -187,19 +187,19 @@ impl<T: From<usize> + PartialEq + Clone> Decode<T> for CheckedValue<T> {
...
             ),
             Type::Struct(members) => CheckedValue::Struct(
                 members
                     .into_iter()
-                    .scan(0, |state, (id, ty)| {
-                        let new_state = *state + ty.get_primitive_count();
-                        let res = CheckedValue::decode(raw[*state..new_state].to_vec(), ty);
+                    .scan(0, |state, member| {
+                        let new_state = *state + member.ty.get_primitive_count();
+                        let res = CheckedValue::decode(raw[*state..new_state].to_vec(), *member.ty);
```
[the remark this answers is not in the document]
Comment (dark64) [reply zokrates_abi/src/lib.rs]: `member.ty` is of type `Box<Type>` but `decode` expects `Type`, so dereferencing is needed here I believe

File: zokrates_core/src/semantics.rs [hunk 3]
```diff
@@ -3369,8 +3374,8 @@ mod tests {
                 .mock();
 
                 let expected_type = Type::Struct(vec![
-                    ("foo".to_string(), Type::FieldElement),
-                    ("bar".to_string(), Type::Boolean),
+                    StructMember::new("foo".to_string(), box Type::FieldElement),
```
[the remark this answers is not in the document]
Comment (dark64) [reply zokrates_core/src/semantics.rs]: Sure, makes sense

File: zokrates_abi/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Schaeff) [reply zokrates_abi/src/lib.rs]: oh right thanks

File: zokrates_core/src/static_analysis/propagation.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Schaeff) [reply zokrates_core/src/static_analysis/propagation.rs]: right my bad, was on mobile..

File: zokrates_core/src/typed_absy/abi.rs [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+pub type AbiOutput = Type;
+
+#[derive(Serialize, Deserialize, Debug, Eq, PartialEq)]
+pub struct Abi {
+    pub inputs: Vec<AbiInput>,
+    pub outputs: Vec<AbiOutput>,
+}
+
+pub trait Generator {
```
Comment (Schaeff) [inline zokrates_core/src/typed_absy/abi.rs]: Can we rename this trait to `Abi`?

File: zokrates_core/src/typed_absy/abi.rs [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+                    ]),
+                },
+            ],
+            outputs: vec![Type::Array(ArrayType::new(Type::FieldElement, 2))],
+        };
+
+        let json = serde_json::to_string(&abi).unwrap();
+        assert_eq!(
+            &json,
+            r#"{"inputs":[{"name":"a","public":false,"type":"field"},{"name":"b","public":true,"type":"struct","components":[{"name":"c","type":"field"},{"name":"d","type":"bool"}]}],"outputs":[{"type":"array","components":{"size":2,"type":"field"}}]}"#
```
Comment (Schaeff) [inline zokrates_core/src/typed_absy/abi.rs]: Could we pretty-print this for clarity?

Comment (Schaeff) [reply zokrates_core/src/typed_absy/abi.rs]: Also let's have a few others like:

```
{"inputs": [], "outputs": []}
{"inputs": [{"name":"a","public":false,"type":"field"}], "outputs": []}
{"inputs": [{"name":"a","public":true,"type":"field"}], "outputs": []}
```

File: zokrates_core/src/typed_absy/abi.rs (hunk 4 above)
Comment (dark64) [reply zokrates_core/src/typed_absy/abi.rs]: yes good point

**Pull request merged**